components/server/src/express/ws-connection-handler.ts

```diff
@@ -30,19 +30,16 @@ export class WsConnectionHandler implements Disposable {
                 try {
                     switch (ws.readyState) {
                         case websocket.CLOSED:
-                            // ws should not be in the clients list anymore, but still happens:
-                            // we rely on a 'close' event being generated, but never receive it. At the same time, the readyState is 'CLOSED' (3).
-                            // judging from the ws source code, this might only happen if an earlier registered handler throws an (unhandled) error.
-                            log.warn("websocket in strange state", { readyState: ws.readyState });
-
-                            // the following is a hack trying to mitigate the effects of leaking CLOSED websockets
-                            if (process.env.EXPERIMENTAL_WS_TERMINATION) {
-                                try {
-                                    (ws as any).emitClose();
-                                    log.warn("websocket (experimental): close emitted");
-                                } catch (err) {
-                                    log.error("websocket (experimental): error on emit('close')", err);
-                                }
+                            // (AT) for unknown reasons the expected `close` event was not emitted, otherwise this websocket would
+                            // no longer be contained in the list of clients iterated over. make sure we release all resources bound
+                            // to this client connection.
+                            const emitClose = (ws as any).emitClose;
+                            if (typeof emitClose === "function") {
+                                this.clients.delete(ws);
+                                emitClose();
+                                log.warn(
+                                    "websocket in strange state. forcefully emitting a close event to release resources.",
```

Comment:
A metric would be better here than a log. Alternatively, enrich it with some data - ie Connection ID to make it useful for tracking down which connection.

Reply:
Reading on https://github.com/websockets/ws/blob/master/lib/websocket.js#L245-L259, we shouldn't even end up in this situation. 

What kind of metric would you suggest? Assuming this is a bug, we want to log it. +1 for logging more details on the connection. Not sure which id is meant, I'll go with full `ClientMetadata`.

Reply:
done

Reply:
Basically it comes down to what is it you're trying to observe in the system. Presumably you're logging it so that you can get an idea that it is happening. For that use-case, a counter metric you add to a dashboard is a better choice - it's cheaper and doesn't require deep context.

If, however, this is just cleanup and we don't really care about how often, or details, of the problem, we might as well not log at all.